src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs

```diff
@@ -304,29 +534,253 @@ private void OnMvcBeforeAction(object arg)
+                {
+                    // Not using endpoint routing
+                    string rawRouteTemplate = actionDescriptor.AttributeRouteInfo?.Template;
+                    RouteTemplate routeTemplate = null;
+                    if (rawRouteTemplate is not null)
+                    {
+                        try
+                        {
+                            routeTemplate = TemplateParser.Parse(rawRouteTemplate);
+                        }
+                        catch { }
+                    }
 
-                string httpMethod = request.Method?.ToUpperInvariant() ?? "UNKNOWN";
-                string routeTemplate = actionDescriptor.AttributeRouteInfo?.Template;
-                if (routeTemplate is null)
+                    if (routeTemplate is null)
+                    {
+                        var routeData = httpContext.Features.Get<IRoutingFeature>()?.RouteData;
+                        if (routeData is not null)
+                        {
+                            var route = routeData.Routers.OfType<RouteBase>().FirstOrDefault();
+                            routeTemplate = route?.ParsedTemplate;
+                        }
+                    }
+
+                    if (routeTemplate is not null)
+                    {
+                        // If we don't have a route, don't overwrite the existing resource name
+                        var resourcePathName = SimplifyRoutePattern(
+                            routeTemplate,
+                            routeValues,
+                            areaName: areaName,
+                            controllerName: controllerName,
+                            actionName: actionName);
+
+                        resourceName = $"{trackingFeature.HttpMethod} {request.PathBase}{resourcePathName}";
+                        aspNetRoute = routeTemplate?.TemplateText.ToLowerInvariant();
+                    }
+                }
+
+                // mirror the parent if we couldn't extract a route
+                span.ResourceName = resourceName ?? parentSpan.ResourceName;
+
+                mvcSpanTags.AspNetCoreAction = actionName;
+                mvcSpanTags.AspNetCoreController = controllerName;
+                mvcSpanTags.AspNetCoreArea = areaName;
+                mvcSpanTags.AspNetCorePage = pagePath;
+                mvcSpanTags.AspNetCoreRoute = aspNetRoute;
+
+                if (!isUsingEndpointRouting && isFirstExecution)
+                {
+                    // If we're using endpoint routing or this is a pipeline re-execution,
+                    // these will already be set correctly
+                    if (parentSpan.Tags is AspNetCoreTags parentTags)
+                    {
+                        parentTags.AspNetCoreRoute = aspNetRoute;
+                    }
+
+                    parentSpan.ResourceName = span.ResourceName;
+                }
+            }
+        }
+
+        private void OnRoutingEndpointMatched(object arg)
```

Comment:
As an ASP.NET Core newbie, assuming each of these private methods are hit in one execution, are the order of these private methods executed in the same order as they are defined in source? It would help readability (for me) if those two orders matched

Reply:
Yeah, great point, now you mention it, this was messing with my head too 😆 done in https://github.com/DataDog/dd-trace-dotnet/pull/1289/commits/2740121e762e09cadef34d593d28b0a8467790b0